### PR TITLE
Convert more btests to btest_exit. NFC

### PR DESCRIPTION
**Pull request by sbc100**
There are still many more to convert.. perhaps we should
be more systematic about it.  For now I'm just doing them
in batches by hand to keep the reviews of reasonable size.

### CODE REVIEW

File: tests/test_browser.py [hunk 1]
```diff
@@ -1805,7 +1805,7 @@ def test_clientside_vertex_arrays_es3(self):
     self.btest('clientside_vertex_arrays_es3.c', reference='gl_triangle.png', args=['-s', 'FULL_ES3=1', '-s', 'USE_GLFW=3', '-lglfw', '-lGLESv2'])
 
   def test_emscripten_api(self):
-    self.btest('emscripten_api_browser.cpp', '1', args=['-s', 'EXPORTED_FUNCTIONS=_main,_third', '-lSDL'])
```
Comment (aheejin) [inline tests/test_browser.py:1808]: Sorry I may not have the sufficient context yet.. In the existing code, some calls to `btest` have the expected value of `1`, while other have `0` or something else, but the new `btest_exit` doesn't contain any. Does deleting these different expected values not matter then? Also, some existing test cases report different values depending on tests using `REPORT_RESULT`, but they are all deleted. These values don't matter and can be deleted safely?

Comment (sbc100) [reply tests/test_browser.py:1808]: `btest_exit` can detect non-zero exit codes but the default and simplest thing to do is to expect an exit code of zero.   What I've done in this change is modified some of the tests to report zero on success and non-zero on failure, at least where it makes sense.  As far as I can tell, and this is where you can help me out, I don't think I've modified the semantics of any of the tests.   They still report success and failures in the same cases. 

For an example of a test where I continue to use the non-zero expectation see `test_emscripten_api_infloop` which as assert_returncode=7.   I can't remember why I didn't change that one in particular, but it shows how btest exit works and how it can be used to assert on non-zero exit codes.

File: tests/emscripten_api_browser.c [hunk 2]
```diff
@@ -105,7 +104,7 @@ int main() {
 
   assert(ratio == ratio2);
 
-  atexit(never); // should never be called - it is wrong to exit the runtime orderly if we have async calls!
+  atexit(check_exit_ok);
```
Comment (aheejin) [inline tests/emscripten_api_browser.c:107]: The previous code says this should never be called.. Is this still not supposed to be called? If so it'd be good to make that clear in the comments. If not what changed?

File: tests/browser/test_emscripten_unwind_to_js_event_loop.c [hunk 3]
```diff
@@ -1,22 +1,17 @@
...
-#ifdef REPORT_RESULT
-	// Should not reach here
-	REPORT_RESULT(-1);
-#endif
+int main() {
+  emscripten_set_timeout(timeout, 2000, 0);
+  emscripten_unwind_to_js_event_loop();
+  printf("This should not be called!\n");
+  // Should not reach here
+  return 99;
```
Comment (aheejin) [inline tests/browser/test_emscripten_unwind_to_js_event_loop.c:16]: Do all these `return 99`s mean this cannot be reached? Would it be better to make `emscripten_unreachable()` or something in the test framework?

File: tests/emscripten_api_browser_infloop.cpp [hunk 4]
```diff
@@ -19,8 +20,8 @@ struct Class {
     printf("waka %d\n", x++);
 
     if (x == 7 || x < 0) {
-      REPORT_RESULT(x);
       emscripten_cancel_main_loop();
+      exit(x);
```
Comment (aheejin) [inline tests/emscripten_api_browser_infloop.cpp:24]: Including this, there are several places the order of `emscripten_cancel_main_loop` and `exit` has changed.. Why is that?

File: tests/emscripten_get_device_pixel_ratio.c [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
-	if (result) {
-		printf("Test succeeded!\n");
-	}
-#ifdef REPORT_RESULT
-	REPORT_RESULT(result);
-#endif
+int main() {
+  double devicePixelRatio = emscripten_get_device_pixel_ratio();
+  printf("window.devicePixelRatio = %f.\n", devicePixelRatio);
+  assert(devicePixelRatio > 0);
```
Comment (aheejin) [inline tests/emscripten_get_device_pixel_ratio.c:8]: Why is this necessary? If the assertion fails here the code below becomes moot..

File: tests/fetch/example_synchronous_fetch.cpp [hunk 6]
```diff
@@ -16,7 +17,6 @@ int main()
   attr.attributes = EMSCRIPTEN_FETCH_LOAD_TO_MEMORY | EMSCRIPTEN_FETCH_SYNCHRONOUS;
   emscripten_fetch_t *fetch = emscripten_fetch(&attr, "gears.png");
   printf("Fetch finished with status %d\n", fetch->status);
-#ifdef REPORT_RESULT
-  REPORT_RESULT(fetch->status);
-#endif
+  assert(fetch->status == 200);
```
Comment (aheejin) [inline tests/fetch/example_synchronous_fetch.cpp:20]: Is the assertion failure the expected way to fail? Other tests use `return 1`, so...

File: tests/pthread/test_pthread_locale.c [hunk 7]
```diff
@@ -52,10 +52,7 @@ int main (int argc, char *argv[])
     pthread_join(thread, (void**)&child_loc);
   }
 
-#ifdef REPORT_RESULT
-  int result = (main_loc == child_loc);
-  REPORT_RESULT(result);
-#endif
 
-  pthread_exit(NULL);
+  assert(main_loc == child_loc);
```
Comment (aheejin) [inline tests/pthread/test_pthread_locale.c:56]: Why is this necessary? Isn't the `return` below gonna signal whether the test succeeds or not anyway?

File: tests/pthread/test_pthread_locale.c [hunk 8]
```diff
@@ -52,10 +52,7 @@ int main (int argc, char *argv[])
     pthread_join(thread, (void**)&child_loc);
   }
 
-#ifdef REPORT_RESULT
-  int result = (main_loc == child_loc);
-  REPORT_RESULT(result);
-#endif
 
-  pthread_exit(NULL);
```
Comment (aheejin) [inline tests/pthread/test_pthread_locale.c:60]: We don't need this anymore?

File: tests/pthread/test_pthread_run_script.cpp [hunk 9]
```diff
@@ -1,19 +1,16 @@
...
-{
-void EMSCRIPTEN_KEEPALIVE FinishTest(int result)
+extern "C" void EMSCRIPTEN_KEEPALIVE FinishTest(int result)
 {
   printf("Test finished, result: %d\n", result);
-#ifdef REPORT_RESULT
-  REPORT_RESULT(result);
-#endif
-}
+  assert(result == 1);
```
Comment (aheejin) [inline tests/pthread/test_pthread_run_script.cpp:12]: Here too, is the assertion failure the expected way of failing?

File: tests/webgl_sample_query.cpp [hunk 10]
```diff
@@ -37,9 +38,7 @@ void getQueryResult()
 
   GL_CALL(glDeleteQueries(1, &sampleQuery));
 
-#ifdef REPORT_RESULT
-  REPORT_RESULT(result);
-#endif
+  exit(result);
```
Comment (aheejin) [inline tests/webgl_sample_query.cpp:41]: Is this `result` expected to be 0 when the test succeeds? The same question for webgl_timer_query.cpp.

**Review by aheejin — COMMENTED** [review]
I see, the removal of random return and expected values looks good! Still some more question though:

File: tests/fetch/example_synchronous_fetch.cpp (hunk 6 above)
Comment (sbc100) [reply tests/fetch/example_synchronous_fetch.cpp:20]: Assertion failure causes the tests to abort which generates a non-zero return code yes.

File: tests/fetch/to_indexeddb.cpp [hunk 11]
```diff
@@ -50,4 +46,5 @@ int main()
   emscripten_fetch_t *fetch = emscripten_fetch(&attr, "gears.png");
   assert(fetch != 0);
   memset(&attr, 0, sizeof(attr)); // emscripten_fetch() must be able to operate without referencing to this structure after the call.
+  return 99;
```
Comment (sbc100) [inline tests/fetch/to_indexeddb.cpp:49]: There is an example of where we actually do return 99 and we expect it not to be reported as an exit (since there is outstanding async work).  If we ever see "failed: 99" will can guess what went wrong.

File: tests/pthread/test_pthread_locale.c (hunk 7 above)
Comment (sbc100) [reply tests/pthread/test_pthread_locale.c:56]: You are right I will just stick to one or the other.

File: tests/pthread/test_pthread_locale.c (hunk 8 above)
Comment (sbc100) [reply tests/pthread/test_pthread_locale.c:60]: Calling pthread_exit from `main` here is just odd..  it might work but it looks pretty out of place.   I don't see why it was ever here.

File: tests/pthread/test_pthread_run_script.cpp (hunk 9 above)
Comment (sbc100) [reply tests/pthread/test_pthread_run_script.cpp:12]: Yes we can use assert to fail best_exit tests test these days.   (I think that is true of btest too but I can't remember for sure).

File: tests/webgl_sample_query.cpp (hunk 10 above)
Comment (sbc100) [reply tests/webgl_sample_query.cpp:41]: One needs to look at the test_browser.py to see what the expected return code of the test is:

```
def test_webgl_timer_query(self):
  ...
  self.btest_exit('webgl_timer_query.cpp', args=cmd)
```

In this case we see no special `assert_returncode` param so the expected result here is zero.

Should I change this to?:

```
assert(result == 0);
exit(0);
```

Or maybe:

```
exit(result == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
```

There a several different ways we could express it I guess, I'm not sure which is the most clear.   Probably the one with the assert?

Comment (sbc100) [reply tests/webgl_sample_query.cpp:41]: Actually I quite like the current version because it keeps this change obvious in that it converted `REPORT_RESULT(XXX)` directly to `exit(XXX)`.    Kind of nice one to one mapping and it means the expectation in the python code stays the same.

**Review by aheejin — APPROVED** [review]
Thanks for the explanation!
